### PR TITLE
Smaato adapter: Rework multi imp support and add adpod support

### PR DESCRIPTION
**Pull request by el-chuck**
Smaatos Ad Server doesn't fully supports multiple impressions per request, creating an individual request per valid bid request is an alternative solution.

Adding adpod support from video endpoint

### CODE REVIEW

File: adapters/smaato/smaatotest/video/multiple-adpods.json [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+                                "maxduration": 30,
+                                "skip": 1,
+                                "protocols": [
+                                    7
+                                ],
+                                "skipmin": 5,
+                                "api": [
+                                    7
+                                ],
+                                "sequence": 1
```
Comment (Enigo) [inline adapters/smaato/smaatotest/video/multiple-adpods.json]: aha it starts with 1 ! I should change on js side, I hv it starting from 0 there

File: adapters/smaato/smaatotest/video/multiple-adpods.json [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+                            "title": "a-title",
+                            "season": "a-season",
+                            "series": "a-series",
+                            "episode": 1,
+                            "len": 900,
+                            "livestream": 1
+                        }
+                    },
+                    "ext": {
+                        "client": "prebid_server_0.2"
```
Comment (Enigo) [inline adapters/smaato/smaatotest/video/multiple-adpods.json]: shouldn't it be 0.3 version ?

File: adapters/smaato/smaatotest/videosupplemental/bad-adbreak-id-request.json [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        },
+        "user": {
+            "ext": {
+                "data": {}
+            }
+        }
+    },
+    "expectedMakeRequestsErrors": [
+        {
+            "value": "Key path not found",
```
Comment (Enigo) [inline adapters/smaato/smaatotest/videosupplemental/bad-adbreak-id-request.json]: shouldn't it be smth more descriptive? like `no adbreakid found`

File: adapters/smaato/smaatotest/videosupplemental/bad-adm-response.json [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+          ],
+          "bidid": "04db8629-179d-4bcd-acce-e54722969006",
+          "cur": "USD"
+        }
+      }
+    }
+  ],
+  "expectedMakeBidsErrors": [
+    {
+      "value": "Invalid ad markup .",
```
Comment (Enigo) [inline adapters/smaato/smaatotest/videosupplemental/bad-adm-response.json:242]: do we know more - like why exactly it isn't valid ?

File: adapters/smaato/smaato.go [hunk 5]
```diff
@@ -214,75 +157,183 @@ func (a *SmaatoAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalR
...
+	return setImp(imp)
 }
 
-func assignBannerSize(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
-	if banner.W != nil && banner.H != nil {
-		return banner, nil
+func setUser(request *openrtb2.BidRequest) error {
+	var err error
+
+	if request.User != nil && request.User.Ext != nil {
```
Comment (Enigo) [inline adapters/smaato/smaato.go:328]: a bit odd that we require both user and user.ext to be present 
I think those two checks can be split

Comment (Enigo) [reply adapters/smaato/smaato.go:328]: okay after looking at test I got it - the config of User is done in ext

File: adapters/smaato/smaato.go [hunk 6]
```diff
@@ -365,3 +452,67 @@ func setBannerDimension(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
...
+
+func filterPodImpressions(imps []openrtb2.Imp) ([]openrtb2.Imp, []error) {
+	var n int
+	errors := make([]error, 0, len(imps))
+	for _, imp := range imps {
+		if imp.Video == nil {
+			errors = append(errors, &errortypes.BadInput{Message: "Invalid MediaType. Smaato only supports Video for AdPod."})
+			continue
+		}
+		imps[n] = imp
```
Comment (Enigo) [inline adapters/smaato/smaato.go:505]: inline `n++` maybe?

File: adapters/smaato/smaatotest/video/multiple-adpods.json (hunk 2 above)
Comment (el-chuck) [reply adapters/smaato/smaatotest/video/multiple-adpods.json]: not sure where you see this?
did you maybe check by commits?

File: adapters/smaato/smaatotest/videosupplemental/bad-adbreak-id-request.json (hunk 3 above)
Comment (el-chuck) [reply adapters/smaato/smaatotest/videosupplemental/bad-adbreak-id-request.json]: will change this and a few other places

File: adapters/smaato/smaatotest/videosupplemental/bad-adm-response.json (hunk 4 above)
Comment (el-chuck) [reply adapters/smaato/smaatotest/videosupplemental/bad-adm-response.json:242]: admarkup does not fit into one of three soma responses for image/richmedia/video

we return the admarkup (which is "" in this case) which should be enough to investigate

File: adapters/smaato/smaato.go (hunk 6 above)
Comment (el-chuck) [reply adapters/smaato/smaato.go:505]: no doable,
n++ is a statement in go and doesn't have a value

File: adapters/smaato/smaatotest/video/multiple-adpods.json (hunk 2 above)
Comment (Enigo) [reply adapters/smaato/smaatotest/video/multiple-adpods.json]: it is now Outdated
but I swear I saw 0.2 version there

**Comment by Enigo** [comment]
if you can add a check for ttl from header that would be awesome!
otherwise, LGTM

**Comment by Enigo** [comment]
LGTM

**Pull request closed without merging**